<center>
<img src="images/ods_stickers.jpg" />
    
## Introducción al Machine Learning

Basado en material de  [Yury Kashnitsky](https://yorko.github.io). Traducido y editado al español por [Ana Georgina Flesia](https://www.linkedin.com/in/georginaflesia/). Este material esta sujeto a los términos y condiciones de la licencia  [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Se permite el uso irrestricto para todo propósito no comercial.

# <center> Ejercicio 1. Análisis exploratorio con Pandas

<img src="images/images/pandas.jpg"  width=50% />


** En esta tarea se tiene que usar Pandas para responder preguntas sobre el  dataset [Adult](https://archive.ics.uci.edu/ml/datasets/Adult). **

Variables y su tipo:
- `age`: continuous.
- `workclass`: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- `fnlwgt`: continuous.
- `education`: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- `education-num`: continuous.
- `marital-status`: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- `occupation`: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- `relationship`: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- `race`: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- `sex`: Female, Male.
- `capital-gain`: continuous.
- `capital-loss`: continuous.
- `hours-per-week`: continuous.
- `native-country`: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- `salary`: >50K,<=50K

In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.max.columns', 100)
# para dibujar gráficos en jupyter notebook
%matplotlib inline 
import matplotlib.pyplot as plt
import seaborn as sns
# no nos gustan los avisos
# se puden comentar las dos lineas si se quiere tener avisos
import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv('/home/floormolina99/Aprendizaje Automatico/adult.data.csv')
data.head(10)

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   
5   37            Private  284582     Masters             14   
6   49            Private  160187         9th              5   
7   52   Self-emp-not-inc  209642     HS-grad              9   
8   31            Private   45781     Masters             14   
9   42            Private  159449   Bachelors             13   

           marital status          occupation    relationship    race  \
0           Never-married        Adm-clerical   Not-in-family   White   
1      Married-civ-spouse     Exec-managerial         Husband   White   
2                Divorced   Handlers-cleaners   Not-in-family   White   
3      Married-civ-spouse   Handlers-cleaners         Husband   Black   
4      Married-civ-spouse      Prof-specialty            Wife   Black   
5      Married-civ-spouse     Exec-managerial            Wife   White   
6   Married-spouse-absent       Other-service   Not-in-family   Black   
7      Married-civ-spouse     Exec-managerial         Husband   White   
8           Never-married      Prof-specialty   Not-in-family   White   
9      Married-civ-spouse     Exec-managerial         Husband   White   

       sex  capital-gain  capital-loss  hours-per-week  native-country  salary  
0     Male          2174             0              40   United-States   <=50K  
1     Male             0             0              13   United-States   <=50K  
2     Male             0             0              40   United-States   <=50K  
3     Male             0             0              40   United-States   <=50K  
4   Female             0             0              40            Cuba   <=50K  
5   Female             0             0              40   United-States   <=50K  
6   Female             0             0              16         Jamaica   <=50K  
7     Male             0             0              45   United-States    >50K  
8   Female         14084             0              50   United-States    >50K  
9     Male          5178             0              40   United-States    >50K

In [3]:
print(data.columns)

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'salary'],
      dtype='object')


In [4]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  salary          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
None


**1. Cuantas mujeres y hombres (variable *sex* ) estan representadas en este dataset?** 

In [5]:
data['sex'].value_counts()

 Male      21790
 Female    10771
Name: sex, dtype: int64

Hay 21790 hombres y 10771 mujeres representados en el dataframe. 

**2. Cual es la edad promedio (variable *age* ) de las mujeres?**

In [6]:
data[data['sex'] == 'Female']['age'].mean()
#No se por qué me tira NaN, pued el codigo deberia correr bien 

nan

In [7]:
f_filter = data['sex']=='Female'
print("%.2f" % data[f_filter]['age'].mean())
#Pruebo de otra forma equivalente pero sigue tirando error

nan


In [8]:
data = data.replace(to_replace='Male',value=0,regex=True)
data = data.replace(to_replace='Female',value=1,regex=True)
data.head()
#Reemplazo las respuestas en la variable "sex" para que en vez de ser strings sean numeros enteros, para ver si asi los reconoce

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital status          occupation    relationship    race  sex  \
0        Never-married        Adm-clerical   Not-in-family   White    0   
1   Married-civ-spouse     Exec-managerial         Husband   White    0   
2             Divorced   Handlers-cleaners   Not-in-family   White    0   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black    0   
4   Married-civ-spouse      Prof-specialty            Wife   Black    1   

   capital-gain  capital-loss  hours-per-week  native-country  salary  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [9]:
data['sex'].value_counts()

0    21790
1    10771
Name: sex, dtype: int64

In [10]:
data[data['sex'] == 1]['age'].mean()

36.85823043357163

In [ ]:
La edad promedio de las mujeres es 36.86 años

**3. Cual es el porcentaje de ciudadanos alemanes (variable *native-country*)?**

In [11]:
data['native-country'].value_counts(normalize=True)

 United-States                 0.895857
 Mexico                        0.019748
 ?                             0.017905
 Philippines                   0.006081
 Germany                       0.004207
 Canada                        0.003716
 Puerto-Rico                   0.003501
 El-Salvador                   0.003255
 India                         0.003071
 Cuba                          0.002918
 England                       0.002764
 Jamaica                       0.002488
 South                         0.002457
 China                         0.002303
 Italy                         0.002242
 Dominican-Republic            0.002150
 Vietnam                       0.002058
 Guatemala                     0.001966
 Japan                         0.001904
 Poland                        0.001843
 Columbia                      0.001812
 Taiwan                        0.001566
 Haiti                         0.001351
 Iran                          0.001321
 Portugal                      0.001136


In [30]:
data['native-country'].value_counts(normalize=True)['Germany']
# No se por qué me da error porque el comando que estoy usando deberia funcionar 

KeyError: 'Germany'

Hay un 0.42% de ciudadanos alemanes

**4. Cuales son la media y la desviacion estandard de la edad para aquellos que ganan mas de 50k por año (variable *salary*) ?**

In [12]:
data = data.replace(to_replace='<=50K',value=0,regex=True)
data = data.replace(to_replace='>50K',value=1,regex=True)
data.head(10)
#Cambio de nuevo los nombres de la variable "salary" porque cuando las variables no son números no me corren bien los codigos

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   
5   37            Private  284582     Masters             14   
6   49            Private  160187         9th              5   
7   52   Self-emp-not-inc  209642     HS-grad              9   
8   31            Private   45781     Masters             14   
9   42            Private  159449   Bachelors             13   

           marital status          occupation    relationship    race  sex  \
0           Never-married        Adm-clerical   Not-in-family   White    0   
1      Married-civ-spouse     Exec-managerial         Husband   White    0   
2                Divorced   Handlers-cleaners   Not-in-family   White    0   
3      Married-civ-spouse   Handlers-cleaners         Husband   Black    0   
4      Married-civ-spouse      Prof-specialty            Wife   Black    1   
5      Married-civ-spouse     Exec-managerial            Wife   White    1   
6   Married-spouse-absent       Other-service   Not-in-family   Black    1   
7      Married-civ-spouse     Exec-managerial         Husband   White    0   
8           Never-married      Prof-specialty   Not-in-family   White    1   
9      Married-civ-spouse     Exec-managerial         Husband   White    0   

   capital-gain  capital-loss  hours-per-week  native-country  salary  
0          2174             0              40   United-States       0  
1             0             0              13   United-States       0  
2             0             0              40   United-States       0  
3             0             0              40   United-States       0  
4             0             0              40            Cuba       0  
5             0             0              40   United-States       0  
6             0             0              16         Jamaica       0  
7             0             0              45   United-States       1  
8         14084             0              50   United-States       1  
9          5178             0              40   United-States       1

In [13]:
data[data['salary'] == 1]['age'].mean()

44.24984058155847

In [14]:
data[data['salary'] == 1]['age'].std()

10.51902771985177

La media de la edad de los que ganan mas de 50K por año es 44.25 años, y la desviación estandar es 10.52 años. 

**5. Cuales son la media y la desviacion estandard de la edad para aquellos que ganan menos de 50k por año (variable *salary*)?**

In [15]:
data[data['salary'] == 0]['age'].mean()

36.78373786407767

In [16]:
data[data['salary'] == 0]['age'].std()

14.020088490824813

In [ ]:
La media de la edad de los que ganan mas de 50K por año es 36.78 años, y la desviación estandar es 14.02 años.

**6. Es cierto que la gente que gana mas que 50K tiene al menos educacion media completa? (variable *education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* or *Doctorate*)**

In [17]:
pd.crosstab(data['education'], data['salary'], normalize=False)

salary            0     1
education                
 10th           871    62
 11th          1115    60
 12th           400    33
 1st-4th        162     6
 5th-6th        317    16
 7th-8th        606    40
 9th            487    27
 Assoc-acdm     802   265
 Assoc-voc     1021   361
 Bachelors     3134  2221
 Doctorate      107   306
 HS-grad       8826  1675
 Masters        764   959
 Preschool       51     0
 Prof-school    153   423
 Some-college  5904  1387

Vemos que la afirmación es falsa, si consideramos tener educación media completa como aquellos que cuentan con un "HS-grad", vemos que hay personas con un nivel de educación más bajo que cobran mas de 50K.

**7. Muestre los estadisticos para cada raza (*race* feature) y cada género (*sex* feature). Use *groupby()* y *describe()*. Encuentre la edad maxima de los hombres de la raza *Amer-Indian-Eskimo*.**

In [18]:
columns_to_show = ['hours-per-week',
                   'capital-gain',
                   'capital-loss',
                   'age'
                  ]
data.groupby(['race'])[columns_to_show].describe(percentiles=[])

hours-per-week                                         \
                             count       mean        std  min   50%   max   
race                                                                        
 Amer-Indian-Eskimo          311.0  40.048232  11.695364  3.0  40.0  84.0   
 Asian-Pac-Islander         1039.0  40.127045  12.556816  1.0  40.0  99.0   
 Black                      3124.0  38.422855  10.315646  1.0  40.0  99.0   
 Other                       271.0  39.468635  11.143755  5.0  40.0  98.0   
 White                     27816.0  40.689100  12.544796  1.0  40.0  99.0   

                    capital-gain                                               \
                           count         mean          std  min  50%      max   
race                                                                            
 Amer-Indian-Eskimo        311.0   625.266881  2753.238961  0.0  0.0  27828.0   
 Asian-Pac-Islander       1039.0  1478.358037  9986.156906  0.0  0.0  99999.0   
 Black                    3124.0   609.940461  5139.653447  0.0  0.0  99999.0   
 Other                     271.0   934.660517  8625.128995  0.0  0.0  99999.0   
 White                   27816.0  1121.660375  7504.533302  0.0  0.0  99999.0   

                    capital-loss                                           \
                           count       mean         std  min  50%     max   
race                                                                        
 Amer-Indian-Eskimo        311.0  34.176849  245.583106  0.0  0.0  1980.0   
 Asian-Pac-Islander       1039.0  97.222329  423.556931  0.0  0.0  2457.0   
 Black                    3124.0  60.385083  337.394121  0.0  0.0  4356.0   
 Other                     271.0  61.070111  322.452705  0.0  0.0  2179.0   
 White                   27816.0  90.806155  410.833347  0.0  0.0  4356.0   

                         age                                          
                       count       mean        std   min   50%   max  
race                                                                  
 Amer-Indian-Eskimo    311.0  37.173633  12.447130  17.0  35.0  82.0  
 Asian-Pac-Islander   1039.0  37.746872  12.825133  17.0  36.0  90.0  
 Black                3124.0  37.767926  12.759290  17.0  36.0  90.0  
 Other                 271.0  33.457565  11.538865  17.0  31.0  77.0  
 White               27816.0  38.769881  13.782306  17.0  37.0  90.0

In [19]:
columns_to_show = ['hours-per-week',
                   'capital-gain',
                   'capital-loss',
                   'age'
                  ]
data.groupby(['sex'])[columns_to_show].describe(percentiles=[])

hours-per-week                                        capital-gain  \
             count       mean        std  min   50%   max        count   
sex                                                                      
0          21790.0  42.428086  12.119755  1.0  40.0  99.0      21790.0   
1          10771.0  36.410361  11.811300  1.0  40.0  99.0      10771.0   

                                                 capital-loss              \
            mean          std  min  50%      max        count        mean   
sex                                                                         
0    1329.370078  8326.312095  0.0  0.0  99999.0      21790.0  100.213309   
1     568.410547  4924.262944  0.0  0.0  99999.0      10771.0   61.187633   

                                       age                                    \
            std  min  50%     max    count       mean        std   min   50%   
sex                                                                            
0    429.763097  0.0  0.0  3770.0  21790.0  39.433547  13.370630  17.0  38.0   
1    340.907518  0.0  0.0  4356.0  10771.0  36.858230  14.013697  17.0  35.0   

           
      max  
sex        
0    90.0  
1    90.0

In [20]:
columns_to_show = ['hours-per-week',
                   'capital-gain',
                   'capital-loss',
                   'age'
                  ]
data.groupby(['race','sex'])[columns_to_show].describe(percentiles=[])

hours-per-week                                         \
                                 count       mean        std  min   50%   max   
race                sex                                                         
 Amer-Indian-Eskimo 0            192.0  42.197917  11.596280  3.0  40.0  84.0   
                    1            119.0  36.579832  11.046509  4.0  40.0  84.0   
 Asian-Pac-Islander 0            693.0  41.468975  12.387563  1.0  40.0  99.0   
                    1            346.0  37.439306  12.479459  1.0  40.0  99.0   
 Black              0           1569.0  39.997451  10.909413  1.0  40.0  99.0   
                    1           1555.0  36.834084   9.419960  2.0  40.0  99.0   
 Other              0            162.0  41.851852  11.084779  5.0  40.0  98.0   
                    1            109.0  35.926606  10.300761  6.0  40.0  65.0   
 White              0          19174.0  42.668822  12.194633  1.0  40.0  99.0   
                    1           8642.0  36.296691  12.190951  1.0  40.0  99.0   

                        capital-gain                                       \
                               count         mean           std  min  50%   
race                sex                                                     
 Amer-Indian-Eskimo 0          192.0   675.260417   2929.745443  0.0  0.0   
                    1          119.0   544.605042   2451.591587  0.0  0.0   
 Asian-Pac-Islander 0          693.0  1827.813853  10947.525528  0.0  0.0   
                    1          346.0   778.436416   7675.228631  0.0  0.0   
 Black              0         1569.0   702.454430   4962.113183  0.0  0.0   
                    1         1555.0   516.593569   5312.749129  0.0  0.0   
 Other              0          162.0  1392.185185  11093.711595  0.0  0.0   
                    1          109.0   254.669725   1317.326460  0.0  0.0   
 White              0        19174.0  1368.674455   8442.830669  0.0  0.0   
                    1         8642.0   573.610391   4763.131649  0.0  0.0   

                                 capital-loss                               \
                             max        count        mean         std  min   
race                sex                                                      
 Amer-Indian-Eskimo 0    27828.0        192.0   46.395833  286.562584  0.0   
                    1    15024.0        119.0   14.462185  157.763811  0.0   
 Asian-Pac-Islander 0    99999.0        693.0  120.373737  472.917697  0.0   
                    1    99999.0        346.0   50.852601  296.529225  0.0   
 Black              0    99999.0       1569.0   75.186106  370.976546  0.0   
                    1    99999.0       1555.0   45.450804  299.099591  0.0   
 Other              0    99999.0        162.0   77.746914  370.986720  0.0   
                    1     7688.0        109.0   36.284404  231.796929  0.0   
 White              0    99999.0      19174.0  102.261343  434.156936  0.0   
                    1    99999.0       8642.0   65.390535  352.330817  0.0   

                                          age                              \
                         50%     max    count       mean        std   min   
race                sex                                                     
 Amer-Indian-Eskimo 0    0.0  1980.0    192.0  37.208333  12.049563  17.0   
                    1    0.0  1721.0    119.0  37.117647  13.114991  17.0   
 Asian-Pac-Islander 0    0.0  2457.0    693.0  39.073593  12.883944  18.0   
                    1    0.0  2258.0    346.0  35.089595  12.300845  17.0   
 Black              0    0.0  2824.0   1569.0  37.682600  12.882612  17.0   
                    1    0.0  4356.0   1555.0  37.854019  12.637197  17.0   
 Other              0    0.0  2179.0    162.0  34.654321  11.355531  17.0   
                    1    0.0  1740.0    109.0  31.678899  11.631599  17.0   
 White              0    0.0  3770.0  19174.0  39.652498  13.436029  17.0   
                    1    0.0 

In [33]:
data[(data['race']=='Amer-Indian-Eskimo') & (data['sex']==0)]['age'].max()
#De nuevo, no se por qué este codigo no corre bien porque debería funcionar

nan

**8. De las siguientes categorías, cual es la que tiene la mayor proporcion de ricos (>50K): married or single men (variable *marital-status*)? Considere como casados  aquellos que tienen *marital-status* comenzando por *Married* (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), el resto son bachelors.**

In [22]:
data.loc[data['marital status'].str.contains("Married")]

age          workclass  fnlwgt    education  education-num  \
1       50   Self-emp-not-inc   83311    Bachelors             13   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
5       37            Private  284582      Masters             14   
6       49            Private  160187          9th              5   
...    ...                ...     ...          ...            ...   
32552   43            Private   84661    Assoc-voc             11   
32554   53            Private  321865      Masters             14   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

               marital status          occupation    relationship    race  \
1          Married-civ-spouse     Exec-managerial         Husband   White   
3          Married-civ-spouse   Handlers-cleaners         Husband   Black   
4          Married-civ-spouse      Prof-specialty            Wife   Black   
5          Married-civ-spouse     Exec-managerial            Wife   White   
6       Married-spouse-absent       Other-service   Not-in-family   Black   
...                       ...                 ...             ...     ...   
32552      Married-civ-spouse               Sales         Husband   White   
32554      Married-civ-spouse     Exec-managerial         Husband   White   
32556      Married-civ-spouse        Tech-support            Wife   White   
32557      Married-civ-spouse   Machine-op-inspct         Husband   White   
32560      Married-civ-spouse     Exec-managerial            Wife   White   

       sex  capital-gain  capital-loss  hours-per-week  native-country  salary  
1        0             0             0              13   United-States       0  
3        0             0             0              40   United-States       0  
4        1             0             0              40            Cuba       0  
5        1             0             0              40   United-States       0  
6        1             0             0              16         Jamaica       0  
...    ...           ...           ...             ...             ...     ...  
32552    0             0             0              45   United-States       0  
32554    0             0             0              40   United-States       1  
32556    1             0             0              38   United-States       0  
32557    0             0             0              40   United-States       1  
32560    1         15024             0              40   United-States       1  

[15417 rows x 15 columns]

In [23]:
data['married-single'] = data['marital status'].apply(lambda x: 'Married' if 'Married' in x else 'Not married')
data.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital status          occupation    relationship    race  sex  \
0        Never-married        Adm-clerical   Not-in-family   White    0   
1   Married-civ-spouse     Exec-managerial         Husband   White    0   
2             Divorced   Handlers-cleaners   Not-in-family   White    0   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black    0   
4   Married-civ-spouse      Prof-specialty            Wife   Black    1   

   capital-gain  capital-loss  hours-per-week  native-country  salary  \
0          2174             0              40   United-States       0   
1             0             0              13   United-States       0   
2             0             0              40   United-States       0   
3             0             0              40   United-States       0   
4             0             0              40            Cuba       0   

  married-single  
0    Not married  
1        Married  
2    Not married  
3        Married  
4        Married

In [24]:
data.groupby(['married-single'])['salary'].value_counts(normalize=True)

married-single  salary
Married         0         0.563080
                1         0.436920
Not married     0         0.935546
                1         0.064454
Name: salary, dtype: float64

Recordando que salary=0 es un salario menor a 50K y salary=1 es un salario mayor a 50K, vemos que la mayor proporción de ricos se encuentran en el grupo de casados.

**9. Cual es el máximo de horas que una persona trabaja por semana? (variable *hours-per-week*)? Cuantas personas trabajan ese numero de horas y cual es el porcentaje entre esas personas que ademas ganan mucho (>50K) ?**

In [25]:
data['hours-per-week'].describe()

count    32561.000000
mean        40.437456
std         12.347429
min          1.000000
25%         40.000000
50%         40.000000
75%         45.000000
max         99.000000
Name: hours-per-week, dtype: float64

El maximo de horas es 99 

In [26]:
data[data['hours-per-week']== 99.].count()

age               85
workclass         85
fnlwgt            85
education         85
education-num     85
marital status    85
occupation        85
relationship      85
race              85
sex               85
capital-gain      85
capital-loss      85
hours-per-week    85
native-country    85
salary            85
married-single    85
dtype: int64

Son 85 las personas que trabajan 99 horas a la semana

In [27]:
data[data['hours-per-week']== 99.]['salary'].value_counts(normalize=True)

0    0.705882
1    0.294118
Name: salary, dtype: float64

El 70,6% aproximadamente gana menos de 50K y el 29.4% aproximadamente gana mas de 50K, o sea solo el 29.4% son ricos

**10. Cuente el numero de horas de trabajo (*hours-per-week*) de aquellos que ganan poco, y de los que ganan mucho (*salary*) para cada pais (*native-country*). Cuales son esos conteos para Japon?**

In [28]:
data.groupby(['salary','native-country'])['hours-per-week'].describe(percentiles=[])

count       mean        std   min   50%   max
salary native-country                                                  
0       ?                 437.0  40.164760  12.817002   1.0  40.0  99.0
        Cambodia           12.0  41.416667   3.088346  40.0  40.0  50.0
        Canada             82.0  37.914634  13.012056   8.0  40.0  70.0
        China              55.0  37.381818  11.439844   1.0  40.0  60.0
        Columbia           57.0  38.684211   8.610952   5.0  40.0  60.0
...                         ...        ...        ...   ...   ...   ...
1       Thailand            3.0  58.333333  14.433757  50.0  50.0  75.0
        Trinadad&Tobago     2.0  40.000000   0.000000  40.0  40.0  40.0
        United-States    7171.0  45.505369  11.025092   1.0  40.0  99.0
        Vietnam             5.0  39.200000   1.788854  36.0  40.0  40.0
        Yugoslavia          6.0  49.500000  11.202678  40.0  46.0  65.0

[82 rows x 6 columns]

In [29]:
data[data['native-country']== 'Japan'].groupby(['salary'])['hours-per-week'].describe(percentiles=[])
#Tampoco se por qué me tira error porque este código deberia correr bien 

ValueError: index must be a MultiIndex to unstack, <class 'pandas.core.indexes.numeric.Int64Index'> was passed